x/exchange/msg.go

```diff
@@ -0,0 +1,464 @@
+		errs = append(errs, fmt.Errorf("invalid market id: cannot be zero"))
+	}
+
+	// Nothing to validate for the AllowUserSettlement field.
+
+	return errors.Join(errs...)
+}
+
+func (m MsgMarketUpdateUserSettleRequest) GetSigners() []sdk.AccAddress {
+	addr := sdk.MustAccAddressFromBech32(m.Admin)
+	return []sdk.AccAddress{addr}
+}
+
+func (m MsgMarketManagePermissionsRequest) ValidateBasic() error {
+	var errs []error
+
+	if _, err := sdk.AccAddressFromBech32(m.Admin); err != nil {
+		errs = append(errs, fmt.Errorf("invalid administrator %q: %w", m.Admin, err))
+	}
+
+	if m.MarketId == 0 {
+		errs = append(errs, fmt.Errorf("invalid market id: cannot be zero"))
+	}
+
+	if m.HasUpdates() {
+		for _, addrStr := range m.RevokeAll {
+			if _, err := sdk.AccAddressFromBech32(addrStr); err != nil {
+				errs = append(errs, fmt.Errorf("invalid revoke-all address %q: %w", addrStr, err))
+			}
+		}
+		if ContainsString(m.RevokeAll, m.Admin) {
+			errs = append(errs, fmt.Errorf("message administrator %s cannot revoke all of their permissions", m.Admin))
+		}
+
+		if err := ValidateAccessGrantsField("to-revoke", m.ToRevoke); err != nil {
+			errs = append(errs, err)
+		}
+
+		toRevokeByAddr := make(map[string]AccessGrant, len(m.ToRevoke))
+		for _, ag := range m.ToRevoke {
+			if ContainsString(m.RevokeAll, ag.Address) {
+				errs = append(errs, fmt.Errorf("address %s appears in both the revoke-all and to-revoke fields", ag.Address))
+			}
+			if ag.Address == m.Admin && ag.Contains(Permission_permissions) {
+				errs = append(errs, fmt.Errorf("message administrator %s cannot revoke their own ability to manage permissions",
+					ag.Address))
+			}
+			toRevokeByAddr[ag.Address] = ag
+		}
+
+		if err := ValidateAccessGrantsField("to-grant", m.ToGrant); err != nil {
+			errs = append(errs, err)
+		}
+
+		for _, ag := range m.ToGrant {
+			toRev, ok := toRevokeByAddr[ag.Address]
+			if ok {
+				for _, perm := range ag.Permissions {
+					if toRev.Contains(perm) {
+						errs = append(errs, fmt.Errorf("address %s has both revoke and grant %q", ag.Address, perm.SimpleString()))
+					}
+				}
+			}
+		}
+	} else {
+		errs = append(errs, errors.New("no updates"))
+	}
+
+	return errors.Join(errs...)
+}
+
+// HasUpdates returns true if this has at least one permission change, false if devoid of updates.
+func (m MsgMarketManagePermissionsRequest) HasUpdates() bool {
+	return len(m.RevokeAll) > 0 || len(m.ToRevoke) > 0 || len(m.ToGrant) > 0
+}
+
+func (m MsgMarketManagePermissionsRequest) GetSigners() []sdk.AccAddress {
+	addr := sdk.MustAccAddressFromBech32(m.Admin)
+	return []sdk.AccAddress{addr}
+}
+
+func (m MsgMarketManageReqAttrsRequest) ValidateBasic() error {
+	var errs []error
+
+	if _, err := sdk.AccAddressFromBech32(m.Admin); err != nil {
+		errs = append(errs, fmt.Errorf("invalid administrator %q: %w", m.Admin, err))
+	}
+
+	if m.MarketId == 0 {
+		errs = append(errs, fmt.Errorf("invalid market id: cannot be zero"))
+	}
+
+	if m.HasUpdates() {
```

Comment:
Returning early means you don't have to parse/scan the rest of the method and is a major reason why the go convention of return on error is used.

This is bordering on bike shedding at this point and I fear we are getting bogged down in details that are not helpful ...

Reply:
From a usability standpoint, it's great when I get a list of all errors rather than just the first error encountered. When you get just one error, you fix it, try again, then get another, and you get bogged down in a tedious fix+try loop. It's like when unit tests stop at the first broken test and don't run any other tests; you're only getting part of the picture. That's why I did it like this for all the message validation, and in a few other places where an error from one thing doesn't mean the next thing is impossible to check, and where painting a better picture would be helpful.

Early returns are good to use when an error is encountered that makes it impossible to move forward. E.g., if `msg.Validate()` returns an error, there's probably nothing else that can be done, so return that error. But inside `Validate()`, it's significantly more user friendly to identify as many issues as possible. There'll be cases where an error from one field means you don't need to check another, but most of the time, that's not the case.